torch/nn/_functions/thnn/activation.py

```diff
@@ -7,54 +7,107 @@
             input,
             output,
             weight,
-            self.num_parameters
+            ctx.num_parameters
         )
-        self.save_for_backward(input, weight)
+        ctx.save_for_backward(input, weight)
         return output
 
-    def backward(self, grad_output):
-        input, weight = self.saved_tensors
-        # TODO: check if requires grad
+    @staticmethod
+    def backward(ctx, grad_output):
+        input, weight = ctx.saved_variables
+        # alternatively, we could recalculate _backend, num_parameters
+        return PReLUBackward.apply(input, weight, grad_output, ctx._backend, ctx.num_parameters)
+
+
+class PReLUBackward(Function):
+    @staticmethod
+    def forward(ctx, input, weight, grad_output, backend, num_parameters):
+        ctx.save_for_backward(input, weight, grad_output)
+        ctx.num_parameters = num_parameters
         grad_input = input.new()
-        self._backend.PReLU_updateGradInput(
-            self._backend.library_state,
+        backend.PReLU_updateGradInput(
+            backend.library_state,
             input,
             grad_output,
             grad_input,
             weight,
-            self.num_parameters
+            num_parameters
         )
 
         buf = weight.new()
         buf2 = weight.new()
         # TODO: this won't have to be zeroed in the future
         grad_weight = weight.new().resize_as_(weight).zero_()
-        self._backend.PReLU_accGradParameters(
-            self._backend.library_state,
+        backend.PReLU_accGradParameters(
+            backend.library_state,
             input,
             grad_output,
             grad_input,
             weight,
             grad_weight,
             buf,
             buf2,
-            self.num_parameters,
+            num_parameters,
             1
         )
         return grad_input, grad_weight
 
+    @staticmethod
+    def backward(ctx, ggI, ggW):
+        input, weight, gO = ctx.saved_variables
+        positive_mask = (input > 0).type_as(ggI)
+        nonpositive_mask = (input <= 0).type_as(ggW)
+        # Explanation: Let input be i, weight be w, grad_output be gO.
+        # f(i, w) = i  if i > 0
+        #         = wi if i <= 0
+        # df/dx * gO  = gO      if i > 0      df/dw * g0 = 0      if i > 0
+        #             = g0 * w  if i <= 0                = g0 * i  if i <= 0
+        # The rest is taking derivatives of these wrt i, w, gO and summing/expanding properly.
+        if ctx.num_parameters == 0:
+            mask = positive_mask + nonpositive_mask * weight.expand_as(input)
+            ggO = ggI * mask + ggW.expand_as(gO) * (nonpositive_mask * input)
+            return ggW.expand_as(gO) * gO * nonpositive_mask, (ggI * gO * nonpositive_mask).sum(), ggO, None, None
+        else:
+            # Expand ggW to match size of ggI; a simple expand doesn't work because
+            # ggW is the size of the input channel (dim==1 unless there is only 1 dimension).
```

Comment:
Am I interpreting this incorrectly?  I was going off the documentation, which says the following:
"Applies element-wise the function \(PReLU(x) = max(0,x) + a * min(0,x)\) Here “a” is a learnable parameter. When called without arguments, nn.PReLU() uses a single parameter “a” across all input channels. If called with nn.PReLU(nChannels), a separate “a” is used for each input channel"

Then, wouldn't gW and ggW be the size of the input channel?